content/en/blog/2024/otel-generative-ai/index.md

```diff
@@ -0,0 +1,204 @@
+[Log events API](https://opentelemetry.io/docs/specs/otel/logs/event-api/)
+specification in the Semantic Conventions for Generative AI. The events API
+allows for us to define specific
+[semantic conventions](https://opentelemetry.io/docs/specs/semconv/general/events/)
+for the user prompts and model responses that we capture. {{% /alert %}}
+
+### Extending Observability with Vendor-Specific Attributes
+
+The Semantic Conventions also define vendor-specific attributes for platforms
+like OpenAI and Azure Inference API, ensuring telemetry captures both general
+and provider-specific details. This added flexibility supports multi-platform
+monitoring and in-depth insights.
+
+## Building the Python Instrumentation Library for OpenAI
+
+This Python-based library for OpenTelemetry captures key telemetry signals for
+OpenAI models, providing developers with an out-of-the-box observability
+solution tailored to AI workloads. The library,
+[hosted within the OpenTelemetry Python Contrib repository](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/opentelemetry-instrumentation-openai-v2%3D%3D2.0b0/instrumentation-genai/opentelemetry-instrumentation-openai-v2),
+automatically collects telemetry from OpenAI model interactions, including
+request and response metadata and token usage.
+
+As generative AI applications grow, additional instrumentation libraries for
+other languages will follow, extending OpenTelemetry support across more tools
+and environments. The current library’s focus on OpenAI highlights its
+popularity and demand within AI development, making it a valuable initial
+implementation.
+
+### Example Usage
+
+Here’s an example of using the OpenTelemetry Python library to monitor a
+generative AI application with the OpenAI client. Make sure you first install
+the library:
+
+```bash
+pip install opentelemetry-instrumentation-openai-v2
+```
+
+Set the following environment variables, updating the endpoint and protocol as
+appropriate:
+
+```bash
+OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
+OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf
+OTEL_SERVICE_NAME=python-opentelemetry-openai
+OPENAI_API_KEY=<replace_with_your_openai_api_key>
+
+# Set to false or remove to disable log events
+OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT=true
+```
+
+Then include the following code in your Python application:
+
+```python
+
+# NOTE: OpenTelemetry Python Log Events APIs is in beta
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk._logs import LoggerProvider
+from opentelemetry.sdk._events import EventLoggerProvider
+from opentelemetry.trace import set_tracer_provider
+from opentelemetry._logs import set_logger_provider
+from opentelemetry._events import set_event_logger_provider
+from opentelemetry.sdk.trace.export import SimpleSpanProcessor
+from opentelemetry.sdk._logs.export import SimpleLogRecordProcessor
+from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
```

Comment:
while grpc is ok I'm not sure if all vendors support that. Also it requires platform deps to build. Should we switch this to http instead?

Reply:
I see now, it is because aspire only supports grpc https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/dashboard/standalone?tabs=bash#configure-opentelemetry-sdk

Reply:
@drewby that link isn't correct afterall, as aspire dashboard also supports http. I suggest changing to http in the final example as that lets you use smaller images like alpine without installing gcc which is needed for the grpc package.

e.g. this works if we use http

```
# Use an alpine image to make the runtime smaller
FROM docker.io/python:3.12.7-alpine3.20
RUN python -m pip install --upgrade pip

COPY /requirements.txt /tmp/requirements.txt
RUN pip install -r /tmp/requirements.txt

COPY main.py /

CMD [ "opentelemetry-instrument", "python", "main.py" ]
```

Reply:
specifically, the following let's you use http on localhost:4318

```bash
docker run --rm -it  -p 18888:18888     -p 4318:18890  --name aspire-dashboard     mcr.microsoft.com/dotnet/aspire-dashboard:9.0
```